File: Frontend/src/services/timetableService.ts
// src/services/timetableService.ts
import axiosInstance from './axiosInstance';

/**
 * TimetableEntry interface represents a timetable entry matching the backend TimeTable model.
 */
export interface TimetableEntry {
  id: number;
  date: string;         // ISO date string (e.g., "2025-04-02")
  startTime: string;    // ISO time string (e.g., "09:00:00")
  endTime: string;      // ISO time string (e.g., "10:30:00")
  teacherId: number;
  courseId: number;
  classroom: string;
}

/**
 * timetableService provides methods for performing CRUD operations on timetable data.
 */
const timetableService = {
  /**
   * Retrieves all timetable entries with optional pagination.
   *
   * @param page - The page number (0-indexed). Defaults to 0.
   * @param size - The number of records per page. Defaults to 100.
   * @returns A promise resolving to an array of TimetableEntry objects.
   *
   * Note: If the backend returns a paginated response, we return response.data.content,
   * otherwise we return the entire response data.
   */
  getAll: async (page: number = 0, size: number = 100): Promise<TimetableEntry[]> => {
    const response = await axiosInstance.get(`/timetable`, {
      params: { page, size },
    });
    return response.data.content ?? response.data;
  },

  /**
   * Retrieves a single timetable entry by its ID.
   *
   * @param id - The timetable entry's ID.
   * @returns A promise resolving to the TimetableEntry.
   */
  getById: async (id: number): Promise<TimetableEntry> => {
    const response = await axiosInstance.get(`/timetable/${id}`);
    return response.data;
  },

  /**
   * Creates a new timetable entry.
   *
   * This operation is restricted to admin users.
   *
   * @param entry - The timetable entry data (excluding the auto-generated id).
   * @returns A promise resolving to the newly created TimetableEntry.
   */
  create: async (entry: Omit<TimetableEntry, 'id'>): Promise<TimetableEntry> => {
    const response = await axiosInstance.post(`/timetable`, entry);
    return response.data;
  },

  /**
   * Updates an existing timetable entry by its ID.
   *
   * @param id - The timetable entry's ID.
   * @param entry - Partial timetable data for update.
   * @returns A promise resolving to the updated TimetableEntry.
   */
  update: async (id: number, entry: Partial<TimetableEntry>): Promise<TimetableEntry> => {
    const response = await axiosInstance.put(`/timetable/${id}`, entry);
    return response.data;
  },

  /**
   * Deletes a timetable entry by its ID.
   *
   * @param id - The timetable entry's ID.
   * @returns A promise that resolves when deletion is complete.
   */
  delete: async (id: number): Promise<void> => {
    await axiosInstance.delete(`/timetable/${id}`);
  },
};

export default timetableService;

File: Service/SMS/src/main/java/com/qslabs/sms/service/TimeTableService.java
package com.qslabs.sms.service;

import com.qslabs.sms.dto.TimeTableDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for managing course timetables.
 * Provides methods for creating, retrieving, updating, and deleting timetable entries.
 */
public interface TimeTableService {

    /**
     * Retrieves all timetable records with pagination.
     *
     * @param pageable object containing page number, size, and sort details
     * @return paginated list of TimeTableDTOs
     */
    Page<TimeTableDTO> getAllTimeTables(Pageable pageable);

    /**
     * Retrieves a specific timetable entry by its ID.
     *
     * @param id the timetable ID
     * @return TimeTableDTO containing timetable details
     */
    TimeTableDTO getTimeTableById(Long id);

    /**
     * Creates a new timetable entry.
     *
     * @param timeTableDTO data required to create a timetable
     * @return the newly created TimeTableDTO
     */
    TimeTableDTO createTimeTable(TimeTableDTO timeTableDTO);

    /**
     * Updates an existing timetable entry by ID.
     *
     * @param id            the ID of the timetable to update
     * @param timeTableDTO  the updated timetable data
     * @return updated TimeTableDTO
     */
    TimeTableDTO updateTimeTable(Long id, TimeTableDTO timeTableDTO);

    /**
     * Deletes a timetable entry by its ID.
     *
     * @param id the ID of the timetable to delete
     * @return true if deletion was successful, false otherwise
     */
    boolean deleteTimeTable(Long id);

    Long getTodayClassCount();
}
